experimental/segmenter/src/rule_segmenter.rs

```diff
@@ -152,24 +152,42 @@ impl<'l, 's, Y: RuleBreakType<'l, 's>> Iterator for RuleBreakIterator<'l, 's, Y>
                 if break_state == NOT_MATCH_RULE {
                     self.iter = previous_iter;
                     self.current_pos_data = previous_pos_data;
-                    return Some(previous_pos_data.unwrap().0);
+                    return Some(self.get_current_position());
                 }
-                return Some(self.current_pos_data.unwrap().0);
+                return Some(self.get_current_position());
             }
 
             if self.is_break_from_table(left_prop, right_prop) {
-                return Some(self.current_pos_data.unwrap().0);
+                return Some(self.get_current_position());
             }
         }
     }
 }
 
 impl<'l, 's, Y: RuleBreakType<'l, 's>> RuleBreakIterator<'l, 's, Y> {
+    pub(crate) fn advance_iter(&mut self) {
+        self.current_pos_data = self.iter.next();
+    }
+
+    pub(crate) fn is_eof(&self) -> bool {
+        self.current_pos_data.is_none()
+    }
+
     pub(crate) fn get_current_break_property(&self) -> u8 {
         self.get_break_property(self.get_current_codepoint())
     }
 
-    fn get_current_codepoint(&self) -> Y::CharType {
+    pub(crate) fn get_current_position(&self) -> usize {
+        debug_assert!(!self.is_eof());
+        #[allow(clippy::expect_used)] // Caller should check is_eof() before calling this
```

Comment:
Nit: Hiding the panic away in a helper function does against the spirit of the style guide. It would be better if this function returned `Option` and the call sites were structured in such a way that `None` values did early returns, exit from a loop, etc.

Reply:
I removed `#[allow(clippy::expect_used)]` in #2817.